File: app/src/main/java/cn/devcxl/photosync/ptp/detect/CameraDetector.java
package cn.devcxl.photosync.ptp.detect;

import android.hardware.usb.UsbDevice;
import android.util.Log;

import java.util.Arrays;
import java.util.List;

/**
 * Created by rainx on 2017/5/23.
 */

public class CameraDetector {

    // Select appropriate deviceInitiator, VIDs see http://www.linux-usb.org/usb.ids

    public static final int VENDOR_ID_CANON = 0x04a9;
    public static final int VENDOR_ID_NIKON = 0x04b0;
    public static final int VENDOR_ID_SONY  = 0x054c;

    public static final int VENDOR_ID_OTHER = 0xffff;

    static List<Integer> vendorIds;
    static  {
        vendorIds = Arrays.asList(
                VENDOR_ID_CANON,
                VENDOR_ID_NIKON,
                VENDOR_ID_SONY
        );
    }


    private UsbDevice device;
    public CameraDetector(UsbDevice device) {
        this.device = device;
    }


    public int getSupportedVendorId() {
        if (vendorIds.contains(device.getVendorId())) {
            return device.getVendorId();
        } else {
            return VENDOR_ID_OTHER;
        }
    }

    public String getDeviceUniqName() {
        String manufacturer = device.getManufacturerName();
        String product = device.getProductName();
        String serial;
        try {
            serial = device.getSerialNumber();
        } catch (SecurityException se) {
            // Permission not granted; fall back to vendorId/productId as a pseudo-serial
            serial = "no-permission-" + device.getVendorId() + ":" + device.getProductId();
            Log.w("CameraDetector", "No permission to read serial number; using fallback", se);
        }
        StringBuilder sb = new StringBuilder();
        sb.append(manufacturer)
                .append("_")
                .append(product)
                .append("_")
                .append(serial);
        return sb.toString();
    }
}

File: app/src/main/java/cn/devcxl/photosync/ptp/detect/CameraDetector.kt
package cn.devcxl.photosync.ptp.detect

import android.hardware.usb.UsbDevice
import timber.log.Timber

/**
 * @author devcxl
 */
class CameraDetector(private val device: UsbDevice) {
    fun getSupportedVendorId(): Int {
        return if (vendorIds.contains(device.vendorId)) {
            device.vendorId
        } else {
            VENDOR_ID_OTHER
        }
    }

    fun getDeviceUniqName(): String {
        val manufacturer = device.manufacturerName
        val product = device.productName
        val serial = try {
            device.serialNumber
        } catch (se: SecurityException) {
            Timber.w(se, "No permission to read serial number; using fallback")
            "no-permission-${device.vendorId}:${device.productId}"
        }
        return "${manufacturer}_${product}_${serial}"
    }

    companion object {
        const val VENDOR_ID_CANON: Int = 0x04a9
        const val VENDOR_ID_NIKON: Int = 0x04b0
        const val VENDOR_ID_SONY: Int = 0x054c

        const val VENDOR_ID_OTHER: Int = 0xffff

        private val vendorIds = listOf(
            VENDOR_ID_CANON,
            VENDOR_ID_NIKON,
            VENDOR_ID_SONY
        )
    }
}
